controller_manager/src/controller_manager.cpp

```diff
@@ -524,42 +604,388 @@ controller_interface::return_type ControllerManager::switch_controller(
+            return controller_interface::return_type::ERROR;
+          }
+          // check if preceding controller will not be deactivated
+          else if (
+            is_controller_active(preceding_ctrl_it->c) &&
+            std::find(stop_request_.begin(), stop_request_.end(), preceding_ctrl_it->info.name) ==
+              stop_request_.end())
+          {
+            RCLCPP_WARN(
+              get_logger(),
+              "Could not deactivate controller with name '%s' because "
+              "preceding controller with name '%s' is active and will not be deactivated.",
+              controller_it->info.name.c_str(), preceding_ctrl_it->info.name.c_str());
+            return controller_interface::return_type::ERROR;
+          }
+          // TODO(destogl): this should be discussed how to it the best - just a placeholder for now
+          // else if (
+          //  strictness ==
+          //  controller_manager_msgs::srv::SwitchController::Request::MANIPULATE_CONTROLLERS_CHAIN)
+          // {
+          // // insert to the begin of start request list to be started before preceding controller
+          //   start_request_.insert(start_request_.begin(), preceding_ctrl_name);
+          // }
+        }
+      }
+      return controller_interface::return_type::OK;
+    };
+
+  for (auto ctrl_it = stop_request_.begin(); ctrl_it != stop_request_.end(); ++ctrl_it)
+  {
+    auto controller_it = std::find_if(
+      controllers.begin(), controllers.end(),
+      std::bind(controller_name_compare, std::placeholders::_1, *ctrl_it));
+    controller_interface::return_type ret = controller_interface::return_type::OK;
+
+    // if controller is not active then skip preceding-controllers checks
+    if (!is_controller_active(controller_it->c))
+    {
+      RCLCPP_WARN(
+        get_logger(), "Controller with name '%s' can not be deactivated since is not active.",
+        controller_it->info.name.c_str());
+      ret = controller_interface::return_type::ERROR;
+    }
+    else
+    {
+      ret = check_preceding_controllers_for_deactivate(
+        controller_it, check_preceding_controllers_for_deactivate);
+    }
+
+    if (ret != controller_interface::return_type::OK)
+    {
+      RCLCPP_WARN(
+        get_logger(),
+        "Could not deactivate controller with name '%s'. (Check above warnings for more details.)"
+        "Check the state of the controllers and their required interfaces using "
+        "`ros2 control list_controllers -v` to get more information.",
+        (*ctrl_it).c_str());
+      if (strictness == controller_manager_msgs::srv::SwitchController::Request::BEST_EFFORT)
+      {
+        stop_request_.erase(ctrl_it--);
+      }
+      if (strictness == controller_manager_msgs::srv::SwitchController::Request::STRICT)
+      {
+        RCLCPP_ERROR(get_logger(), "Aborting, no controller is switched! (::STRICT switch)");
+        // reset all lists
+        stop_request_.clear();
```

Comment:
this pattern of "zero everything" repeats multiple times, one could write a function `clear_requests()` to do this

Reply:
done.